ruby/version.go

```diff
@@ -0,0 +1,162 @@
+package ruby
+
+import (
+	"errors"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+var (
+	anchoredVersion = regexp.MustCompile(`^\s*([0-9]+(\.[0-9a-zA-Z]+)*(-[0-9A-Za-z-]+(\.[0-9A-Za-z-]+)*)?)?\s*$`)
+	segment         = regexp.MustCompile(`\d+|[A-Za-z]+`)
+
+	errInvalidVersion = errors.New("invalid gem version")
+)
+
+type Segment interface {
+	Compare(other Segment) int
+}
+
+var (
+	_ Segment = (*stringSegment)(nil)
+	_ Segment = (*numericSegment)(nil)
+)
+
+type stringSegment string
+
+func (s stringSegment) Compare(other Segment) int {
+	switch seg := other.(type) {
+	case numericSegment:
+		return -1
+	case stringSegment:
+		return strings.Compare(string(s), string(seg))
+	default:
+		panic("Programmer error")
+	}
+}
+
+type numericSegment uint64
+
+func (n numericSegment) Compare(other Segment) int {
+	switch seg := other.(type) {
+	case stringSegment:
+		return +1
+	case numericSegment:
+		if n == seg {
+			return 0
+		}
+		if n < seg {
+			return -1
+		}
+		return +1
+	default:
+		panic("Programmer error")
+	}
+}
+
+// Version is a RubyGem version.
+// This is based on https://github.com/rubygems/rubygems/blob/v3.4.21/lib/rubygems/version.rb.
+type Version struct {
+	v          string
+	prerelease bool
+	segs       []Segment
+}
+
+func NewVersion(version string) (Version, error) {
+	if !anchoredVersion.MatchString(version) {
+		return Version{}, errInvalidVersion
+	}
+
+	version = strings.TrimSpace(version)
+	if version == "" {
+		version = "0"
+	}
+	version = strings.ReplaceAll(version, `-`, `.pre.`)
+
+	v := Version{
+		v: version,
+	}
+	v.canonicalSegments()
+
+	return v, nil
+}
+
+func (v Version) Compare(other Version) int {
+	segs := v.segs
+	otherSegs := other.segs
+
+	leftLen := len(segs)
+	rightLen := len(otherSegs)
+	limit := max(leftLen, rightLen)
+	for i := 0; i < limit; i++ {
+		left, right := Segment(numericSegment(0)), Segment(numericSegment(0))
+		if i < leftLen {
+			left = segs[i]
+		}
+		if i < rightLen {
+			right = otherSegs[i]
+		}
+
+		if cmp := left.Compare(right); cmp != 0 {
+			return cmp
+		}
+	}
+
+	return 0
+}
+
+func (v *Version) canonicalSegments() {
+	v.partitionSegments()
+
+	// Remove trailing zero segments.
+	i := len(v.segs) - 1
+	for ; i >= 0; i-- {
+		seg, ok := v.segs[i].(numericSegment)
+		if !ok || seg != 0 {
+			break
+		}
+	}
+	v.segs = v.segs[:i+1]
+
+	// Remove all zero segments preceding the first letter in a prerelease version.
+	if v.prerelease {
+		var start *int
```

Comment:
Note: this does not account for 1.0.0.alpha == 1.alpha at this time, so need to go over this again